File: apps/api/app/routes/tax_calc.py
"""自賠責・重量税計算 API"""
from __future__ import annotations

from typing import List

from fastapi import APIRouter
from pydantic import BaseModel, Field

from app.services.tax_calculator import (
    VEHICLE_TYPE_LABELS,
    ECO_TYPE_LABELS,
    calculate,
)

router = APIRouter(prefix="/tax", tags=["tax"])


# ─── Schemas ─────────────────────────────────────────────────────────────────

class TaxCalcRequest(BaseModel):
    vehicle_type: str = Field(..., description="passenger/kei/kei_business/bike_small/moped")
    weight_kg: float = Field(..., ge=0, description="車両重量(kg)")
    first_reg_year: int = Field(..., ge=1950, le=2100)
    first_reg_month: int = Field(..., ge=1, le=12)
    eco_type: str = Field("non_eco", description="non_eco/exempt/eco_75/eco_50/eco_25")
    jibaiseki_months: int = Field(25, description="自賠責加入期間(月)")
    inspection_years: int = Field(2, ge=1, le=2, description="車検期間(1 or 2年)")


class TaxCalcResponse(BaseModel):
    jibaiseki: int
    jyuryozei: int
    total: int
    vehicle_age: int
    age_category: str
    notes: List[str]


class VehicleTypeItem(BaseModel):
    value: str
    label: str


# ─── Endpoints ───────────────────────────────────────────────────────────────

@router.post("/calculate", response_model=TaxCalcResponse)
def calculate_tax(body: TaxCalcRequest) -> TaxCalcResponse:
    result = calculate(
        vehicle_type=body.vehicle_type,
        weight_kg=body.weight_kg,
        first_reg_year=body.first_reg_year,
        first_reg_month=body.first_reg_month,
        eco_type=body.eco_type,
        jibaiseki_months=body.jibaiseki_months,
        inspection_years=body.inspection_years,
    )
    return TaxCalcResponse(**result)


@router.get("/vehicle-types", response_model=List[VehicleTypeItem])
def get_vehicle_types() -> List[VehicleTypeItem]:
    return [
        VehicleTypeItem(value=k, label=v)
        for k, v in VEHICLE_TYPE_LABELS.items()
    ]


@router.get("/eco-types", response_model=List[VehicleTypeItem])
def get_eco_types() -> List[VehicleTypeItem]:
    return [
        VehicleTypeItem(value=k, label=v)
        for k, v in ECO_TYPE_LABELS.items()
    ]

File: apps/web/src/lib/api/taxCalc.ts
import { apiFetch } from "./core";

// ─── Types ───────────────────────────────────────────────────────────────────

export interface TaxCalcRequest {
  vehicle_type: string;
  weight_kg: number;
  first_reg_year: number;
  first_reg_month: number;
  eco_type: string;
  jibaiseki_months: number;
  inspection_years: number;
}

export interface TaxCalcResult {
  jibaiseki: number;
  jyuryozei: number;
  total: number;
  vehicle_age: number;
  age_category: string;
  notes: string[];
}

export interface TaxTypeItem {
  value: string;
  label: string;
}

// ─── Constants ───────────────────────────────────────────────────────────────

export const VEHICLE_TYPES: TaxTypeItem[] = [
  { value: "passenger",    label: "自家用乗用車" },
  { value: "kei",          label: "軽自動車（自家用）" },
  { value: "kei_business", label: "軽自動車（営業用）" },
  { value: "bike_small",   label: "小型二輪（250cc超）" },
  { value: "moped",        label: "原付（125cc以下）" },
];

export const ECO_TYPES: TaxTypeItem[] = [
  { value: "non_eco", label: "非エコカー" },
  { value: "exempt",  label: "エコカー免税" },
  { value: "eco_75",  label: "エコカー75%減" },
  { value: "eco_50",  label: "エコカー50%減" },
  { value: "eco_25",  label: "エコカー25%減" },
];

export const JIBAISEKI_MONTHS_OPTIONS = [
  { value: 12, label: "12ヶ月" },
  { value: 24, label: "24ヶ月" },
  { value: 25, label: "25ヶ月" },
  { value: 36, label: "36ヶ月" },
  { value: 37, label: "37ヶ月" },
];

// ─── API Functions ────────────────────────────────────────────────────────────

export async function calculateTax(req: TaxCalcRequest): Promise<TaxCalcResult> {
  return apiFetch<TaxCalcResult>("/api/v1/tax/calculate", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(req),
  });
}
